source/validate.cpp

```diff
@@ -520,24 +538,117 @@ spv_result_t ModuleLayoutPass(ValidationState_t& _,
               inst->words + inst->operands[2].offset;
           if (*storage_class != SpvStorageClassFunction)
             return _.diag(SPV_ERROR_INVALID_LAYOUT)
-                   << "All OpVariable instructions in a function must have a "
-                      "storage class of Function[7]";
+                   << "All Variable instructions in a function must have a "
+                      "storage class of function[7]";
+        } break;
+        default:
           break;
+      }
+      if (_.getLayoutStage() == kLayoutFunctionDeclarations) {
+        switch (opcode) {
+          case SpvOpFunction:
+            if (_.in_function_body()) {
+              return _.diag(SPV_ERROR_INVALID_LAYOUT)
+                     << "Cannot declare a function in a function body";
+            }
+            spvCheckReturn(_.get_functions().RegisterFunction(
+                inst->result_id, inst->type_id,
+                inst->words[inst->operands[2].offset],
+                inst->words[inst->operands[3].offset]));
+            break;
+          case SpvOpFunctionParameter:
+            if (_.in_function_body() == false) {
+              return _.diag(SPV_ERROR_INVALID_LAYOUT) << "Function parameter "
+                                                         "instructions must be "
+                                                         "in a function body";
+            }
+            spvCheckReturn(_.get_functions().RegisterFunctionParameter(
+                inst->result_id, inst->type_id));
+            break;
+          case SpvOpLine:  // ??
+            break;
+          case SpvOpLabel:
```

Comment:
The `CfgPass` function is defined in line 606. This function is only executed when the `getLayoutStage` returns `kFunctionDefinitions`. This stage switch case is skipped in that case (see line 547).

I decided to put those checks in the CfgPass because only function definitions can have blocks. It seemed to be the natural place for something like this. Although now that I think about it I can probably just as easily have that logic here. I can put the layout logic here if you feel this is a better location for something like that.

Anyway line 624 ensures you cannot have function declarations intermingled with function definitions by making sure that the function has at least one block.